File: backend/app/services/meal_service.py
# backend/app/services/meal_service.py
from datetime import date, timedelta
from decimal import Decimal
from typing import List
from uuid import UUID

from ..schemas.meal import MealCreate, MealUpdate, MealOut, DailyTotals
from ..services.supabase_client import supabase

TABLE = "meals"
VIEW_DAILY_TOTALS = "daily_nutrition_totals"

# ---------- CRUD ----------

async def create_meal(data: MealCreate) -> MealOut:
    res = supabase.table(TABLE).insert(data.model_dump(mode="json")).execute()
    if not res.data:
        raise Exception(res.error)
    return MealOut(**res.data[0])

async def update_meal(meal_id: UUID, data: MealUpdate) -> MealOut:
    res = supabase.table(TABLE).update(data.model_dump(exclude_none=True)).eq("id", str(meal_id)).execute()
    if not res.data:
        raise Exception(res.error)
    return MealOut(**res.data[0])

async def delete_meal(meal_id: UUID) -> None:
    res = supabase.table(TABLE).delete().eq("id", str(meal_id)).execute()
    if not res.data:
        raise Exception(res.error)

async def get_meals_by_user_and_date(user_id: UUID, target_date: date) -> List[MealOut]:
    res = supabase.table(TABLE).select("*").eq("user_id", str(user_id)).eq("date", target_date.isoformat()).execute()
    if not res.data:
        raise Exception(res.error)
    return [MealOut(**row) for row in res.data]

# ---------- Aggregation ----------

async def get_daily_totals(user_id: UUID, start_date: date, end_date: date) -> List[DailyTotals]:
    res = (
        supabase.table(VIEW_DAILY_TOTALS)
        .select("*")
        .eq("user_id", str(user_id))
        .gte("date", start_date.isoformat())
        .lte("date", end_date.isoformat())
        .execute()
    )
    if not res.data:
        raise Exception(res.error)
    return [DailyTotals(**row) for row in res.data]

async def get_rolling_averages(user_id: UUID, window_days: int) -> dict:
    end_date = date.today()
    start_date = end_date - timedelta(days=window_days - 1)
    daily = await get_daily_totals(user_id, start_date, end_date)

    if not daily:
        return {
            "calories_avg": 0,
            "protein_avg": 0,
            "carbs_avg": 0,
            "fats_avg": 0,
        }

    n = len(daily)
    return {
        "calories_avg": float(sum(d.total_calories for d in daily) / Decimal(n)),
        "protein_avg": float(sum(d.total_protein_g for d in daily) / Decimal(n)),
        "carbs_avg": float(sum(d.total_carbs_g for d in daily) / Decimal(n)),
        "fats_avg": float(sum(d.total_fats_g for d in daily) / Decimal(n)),
    }

File: frontend/src/services/mealService.ts
// frontend/src/services/mealService.ts
import api from "./api";

export interface MealCreate {
  meal_type: "breakfast" | "lunch" | "dinner" | "snack";
  food_items: Record<string, any>[] | string[];
  calories: number;
  protein_g: number;
  carbs_g: number;
  fats_g: number;
  date: string; // ISO date string
}

export interface MealUpdate {
  meal_type?: "breakfast" | "lunch" | "dinner" | "snack";
  food_items?: Record<string, any>[] | string[];
  calories?: number;
  protein_g?: number;
  carbs_g?: number;
  fats_g?: number;
  date?: string;
}

export interface MealOut extends MealCreate {
  id: string;
  created_at: string;
  updated_at: string;
}

export interface DailyTotals {
  date: string;
  total_calories: number;
  total_protein_g: number;
  total_carbs_g: number;
  total_fats_g: number;
}

export interface RollingAveragesResponse {
  dates: string;
  avg_calories: number;
  avg_protein_g: number;
  avg_carbs_g: number;
  avg_fats_g: number;
}

/**
 * Create a new meal entry
 */
export const createMeal = async (
  token: string,
  meal: MealCreate
): Promise<MealOut> => {
  const res = await api.post<MealOut>("/api/meals/", meal, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return res.data;
};

/**
 * Update an existing meal
 */
export const updateMeal = async (
  token: string,
  mealId: string,
  updateData: MealUpdate
): Promise<MealOut> => {
  const res = await api.put<MealOut>(`/api/meals/${mealId}`, updateData, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return res.data;
};

/**
 * Delete a meal
 */
export const deleteMeal = async (token: string, mealId: string): Promise<void> => {
  await api.delete(`/api/meals/${mealId}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
};

/**
 * Fetch all meals for a user on a given date
 */
export const getMealsByDate = async (
  token: string,
  date: string
): Promise<MealOut[]> => {
  const res = await api.get<MealOut[]>(`/api/meals`, {
    headers: { Authorization: `Bearer ${token}` },
    params: { date: date },
  });
  return res.data;
};

/**
 * Get daily totals between two dates
 */
export const getDailyTotals = async (
  token: string,
  startDate: string,
  endDate: string
): Promise<DailyTotals[]> => {
  const res = await api.get<DailyTotals[]>(
    `api/nutrition/daily-totals`,
    {
      headers: { Authorization: `Bearer ${token}` },
      params: { start: startDate, end: endDate },
    }
  );
  return res.data;
};

/**
 * Get rolling averages for calories/macros
 */
export const getRollingAverages = async (
  token: string,
  days: number = 7
): Promise<RollingAveragesResponse> => {
  const res = await api.get<RollingAveragesResponse>(
    `api/nutrition/rolling-averages`,
    {
      headers: { Authorization: `Bearer ${token}` },
      params: { window: days },
    }
  );
  console.log("Rolling averages response:", res.data);
  return res.data;
};

export default {
  createMeal,
  updateMeal,
  deleteMeal,
  getMealsByDate,
  getDailyTotals,
  getRollingAverages,
};
